c2c: checkpoint cutover progress

msbutler opened the pull request:
This patch adds checkpointing to c2c cutover, which means that if the c2c job
retries on a non terminal error (e.g. a pause or node restart), the cutover
process can resume from the progress it persisted, rather than starting all
over.

This patch also fixes two bugs in the cutoverProgressTracker:
- its `getRangeCount(remainingSpans)` helper used to always return the number
  ranges in the original span to cutover, instead of the ranges that still
needed to be cutover.
- the fractionProgressed would reset to 0 when the cutover process retried. Now
  the fraction should never decrease.

Fixes https://github.com/cockroachdb/cockroach/issues/97384

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105376)
<!-- Reviewable:end -->

chengxiong-ruan commented:
heh, why `O-community` was added?

msbutler commented:
ah, probably because i forgot to add the dr tag

msbutler commented:
er nvm, no idea :D

msbutler committed:
c2c: checkpoint cutover progress

This patch adds checkpointing to c2c cutover, which means that if the c2c job
retries on a non terminal error (e.g. a pause or node restart), the cutover
process can resume from the progress it persisted, rather than starting all
over.

This patch also fixes two bugs in the cutoverProgressTracker:
- its `getRangeCount(remainingSpans)` helper used to always return the number
  ranges in the original span to cutover, instead of the ranges that still
needed to be cutover.
- the fractionProgressed would reset to 0 when the cutover process retried. Now
  the fraction should never decrease.

Fixes #97384

Release note: None

msbutler commented:
stress-race fails on my new test due to https://github.com/cockroachdb/cockroach/issues/104838

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go:
```diff
@@ -529,3 +530,83 @@ func TestCutoverFractionProgressed(t *testing.T) {
+			if updateCount == 3 {
+				close(progressUpdated)
+				// To ensure the job machinery pauses the job before the next update
+				time.Sleep(1 * time.Second)
```
IIUC if this time is not enough then the test will flake? because we will unpause and finish the cutover? I think it means it will flake..
why not just wait on a channel signaled after pause is completed? I guess in the next call to the hook?

if you're keeping this as it, try it under stress+race?

msbutler replied:
good point, it seems to pass stress-race but i changed this to channel just to be safe.

msbutler replied:
good point, it seems to pass stress-race but i changed this to a channel just to be safe.

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go:
```diff
@@ -529,3 +530,83 @@ func TestCutoverFractionProgressed(t *testing.T) {
+	c.DestSysSQL.Exec(t, `RESUME JOB $1`, &replicationJobID)
+	jobutils.WaitForJobToSucceed(t, c.DestSysSQL, replicationJobID)
+
+	// Ensure no spans are left to cutover. Stringify during comparison because
+	// the empty remainingSpans are encoded as
+	// roachpb.Spans{roachpb.Span{Key:/Min, EndKey:/Min}.
```
thanks for this comment :)

lidorcarmel reviewed (commented):
nice!

msbutler commented on pkg/jobs/jobspb/jobs.proto:
```diff
@@ -173,6 +173,12 @@ message StreamIngestionProgress {
 
   reserved 3;
+
+  // RemainingCutoverSpans contains the spans the still need to be cutover once
```
Done.

msbutler replied:
Done.

msbutler reviewed (commented):
thanks for the feedback!

msbutler commented:
added do-not-merge label to remind myself to squash the extra "address lidor comments" commit once you approve

lidorcarmel commented on pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go:
```diff
@@ -480,9 +480,10 @@ func maybeRevertToCutoverTimestamp(
-		spansToRevert         roachpb.Spans
+		shouldRevertToCutover  bool
+		cutoverTimestamp       hlc.Timestamp
+		originalSpansToRevert  roachpb.Spans
```
the original can be a single span, right? the tenant span, so it will never have multiple spans. probably better to keep as a single span so that the reader would know better what's going on?

msbutler replied:
Done.

lidorcarmel commented:
if you don't mind, can you fix the test that should have caught the bug that you found? I think it's https://github.com/cockroachdb/cockroach/blob/327947fb87028f068093726a606b04057d33146f/pkg/ccl/streamingccl/streamingest/stream_ingestion_job_test.go#L503-L515

msbutler commented:
fixed the test as well!

msbutler commented:
Tftr!

bors r=lidorcarmel

msbutler commented:
blathers backport 23.1
